modules/core/src/main/scala/parser.scala

```diff
@@ -312,71 +316,145 @@ object GraphQLParser {
+    (char('#') *> many((charWhere(c => c != '\n' && c != '\r'))) <* charIn('\n', '\r') <* skipWhitespace).void.withContext("comment")
   }
 
   /** Turns a parser into one that skips trailing whitespace and comments */
   def token[A](p: Parser[A]): Parser[A] =
-    p <~ skipWhitespace <~ many(comment)
+    p <* skipWhitespace <* many(comment)
+
+  def token0[A](p: Parser0[A]): Parser0[A] =
+    p <* skipWhitespace <* many(comment)
 
   /**
    * Consumes `left` and `right`, including the trailing and preceding whitespace,
    * respectively, and returns the value of `p`.
    */
-  private def _bracket[A,B,C](left: Parser[B], p: => Parser[A], right: => Parser[C]): Parser[A] =
-    token(left) ~> token(p) <~ token(right)
+  private def _bracket[A,B,C](left: Parser[B], p: => Parser0[A], right: => Parser[C]): Parser[A] =
+    token(left) *> token0(p) <* token(right)
 
   /** Turns a parser into one that consumes surrounding parentheses `()` */
-  def parens[A](p: => Parser[A]): Parser[A] =
-    _bracket(char('('), p, char(')')).named(s"parens(${p.toString})")
+  def parens[A](p: => Parser0[A]): Parser[A] =
+    _bracket(char('('), p, char(')')).withContext(s"parens(${p.toString})")
 
   /** Turns a parser into one that consumes surrounding curly braces `{}` */
-  def braces[A](p: => Parser[A]): Parser[A] = {
+  def braces[A](p: => Parser0[A]): Parser[A] = {
     lazy val q = p
-    _bracket(char('{'), q, char('}')).named(s"braces(${q.toString})")
+    _bracket(char('{'), q, char('}')).withContext(s"braces(${q.toString})")
   }
 
   /** Turns a parser into one that consumes surrounding square brackets `[]` */
-  def squareBrackets[A](p: => Parser[A]): Parser[A] = {
+  def squareBrackets[A](p: => Parser0[A]): Parser[A] = {
     lazy val q = p
-    _bracket(char('['), q, char(']')).named(s"squareBrackets(${q.toString})")
+    _bracket(char('['), q, char(']')).withContext(s"squareBrackets(${q.toString})")
+  }
+}
+
+object Literals {
+
+  lazy val stringLiteral: Parser[String] = {
+
+    lazy val stringCharacter: Parser[String] = (
+      (peek(not(charIn('"', '\\') | lineTerminator)).with1 *> sourceCharacter) |
+        (string("\\u") ~ escapedUnicode).string |
+        (char('\\') ~ escapedCharacter).string
+    )
+
+    lazy val blockStringCharacter = string("TODO").string
+
+    //TODO should this be a Parser[Char] which converts to unicode? I guess so?
+    //See Atto for reference
+    lazy val escapedUnicode: Parser[String] = (digit | hexdig).repExactlyAs[String](4)
+
+    lazy val escapedCharacter = charIn('"', '\\', '/', 'b', 'f', 'n', 'r', 't')
+
+    lazy val sourceCharacter: Parser[String] = (charIn(0x0009.toChar, 0x000A.toChar, 0x000D.toChar) | charIn(0x0020.toChar to 0xFFFF.toChar)).string
+
+    lazy val lineTerminator = (lf | cr | crlf).string
+
+    stringCharacter.rep0.string.with1.surroundedBy(char('"')) | (blockStringCharacter.rep0.string.with1.surroundedBy(string("\"\"\"")))
+
```

Comment:
you can compare this approach with some code I wrote here:

https://github.com/typelevel/cats-parse/blob/85f3c7f3ee67f9fe96c7754bd06e81a00d533da1/bench/src/main/scala/cats/parse/bench/self.scala#L130

this code was designed for a python like language that used either `'` or `"` as the start of your strings and can be simplified somewhat if you have a single approach and set of escapes.